Add defaultFontSizes theme.json (theme.json v3)

ajlende opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## Dev Note 📝 

See https://github.com/WordPress/gutenberg/pull/58409#issuecomment-2078077477

## What?
<!-- In a few words, what is the PR actually doing? -->

Reimplementation of #56661 with a proper theme.json migration.

Adds a `defaultFontSizes` option to disable showing the default font sizes.

When `defaultFontSizes === false` the options are removed from the UI.

In theme.json v3, by default `defaultFontSizes` is `true`.

In theme.json v2, `defaultFontSizes` isn't a valid key, but is treated as `false` where it would be used when a theme has font sizes defined for backwards compatibility of generated CSS.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Having an option for `defaultFontSizes` would give themes more control over which options are available. This option brings consistency to font sizes to match colors, gradients, duotone filters, and shadow presets.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Goals are to migrate the behavior of the font sizes presets to be like other presets that can be disabled via `default*` settings.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

Screenshots below contain example modified `fontSizes` JSON for TT4 to cover various cases of having font sizes and overriding default ones.

For each modified theme.json test that each step works as expected (see screenshots):

1. Test with theme.json `"version": 2`. (trunk and this PR should be the same.)
2. Update your theme.json to `"version": 3`. (Don't forget to update the style variations too.)
3. Update your theme.json (again) with `settings.typography.defaultFontSizes` to `false`.

Then test custom origin styles migration.

1. Start on trunk with a `"version": 2` theme.json with no associated `wp_global_styles` `post_type`s in the DB.
2. Make changes to the global styles UI such as adding a new color.
3. See the `"version": 2` entry in the DB.
4. Switch to this branch.
5. Make more changes to the global styles UI such removing the color.
6. See that the entry is updated to `"version": 3` in the DB.
7. See that `"defaultFontSizes": false` is _not_ added to the entry in the DB.
8. Update the theme's theme.json to `"version": 3`.
9. Make more updates in the UI.
10. See that `"defaultFontSizes": false` is still not added.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

N/A

## Screenshots or screencast <!-- if applicable -->

Each set of screenshots shows a different example theme.json modification to TT4.

- v2 trunk: Screenshot of previous behavior in v2 theme.json.
- v2: Screenshot from this branch for v2 theme.json.
- v3 (default): Screenshot with v3 theme.json. (Should be the same as `"defaultFontSizes": true`.)
- v3 `false`: Screenshot with `"defaultFontSizes": false` set in v3 theme.json.

### TT4 Footnote Override

<details>
<summary>Example JSON</summary>

```jsonc
// Override fontSizes
// Same as TT4 with small renamed to footnote
{
	"settings": {
		"typography": {
			"fontSizes": [
				{
					"fluid": false,
					"name": "Footnote",
					"size": "0.9rem",
					"slug": "footnote"
				},
				{
					"fluid": false,
					"name": "Medium",
					"size": "1.05rem",
					"slug": "medium"
				},
				{
					"fluid": {
						"min": "1.39rem",
						"max": "1.85rem"
					},
					"name": "Large",
					"size": "1.85rem",
					"slug": "large"
				},
				{
					"fluid": {
						"min": "1.85rem",
						"max": "2.5rem"
					},
					"name": "Extra Large",
					"size": "2.5rem",
					"slug": "x-large"
				},
				{
					"fluid": {
						"min": "2.5rem",
						"max": "3.27rem"
					},
					"name": "Extra Extra Large",
					"size": "3.27rem",
					"slug": "xx-large"
				}
			]
		}
	}
}
```

</details>

| v2 trunk | v2 | v3 (default) | v3 `false` |
| -------- | -- | ------------ | ---------- |
| ![font-sizes-v2-override-before](https://github.com/WordPress/gutenberg/assets/5129775/a0360067-87f1-4b0d-b489-275eec6fb969) | ![font-sizes-v2-override-after](https://github.com/WordPress/gutenberg/assets/5129775/7a52e2cd-dbd7-4e13-a284-43f2fa351433) | ![font-sizes-v3-override-true](https://github.com/WordPress/gutenberg/assets/5129775/af80a3fe-7926-4a10-85e5-85bca5406eb9) | ![font-sizes-v3-override-false](https://github.com/WordPress/gutenberg/assets/5129775/46dfb059-6fdf-4560-b3b5-fd645094884d) |

### TT4 Custom Only

<details>
<summary>Example JSON</summary>

```jsonc
// Only custom fontSizes
// Same as override with only footnote and extra-extra-large
{
	"settings": {
		"typography": {
			"fontSizes": [
				{
					"fluid": false,
					"name": "Footnote",
					"size": "0.9rem",
					"slug": "footnote"
				},
				{
					"fluid": {
						"min": "2.5rem",
						"max": "3.27rem"
					},
					"name": "Extra Extra Large",
					"size": "3.27rem",
					"slug": "xx-large"
				}
			]
		}
	}
}
```

</details>

| v2 trunk | v2 | v3 (default) | v3 `false` |
| -------- | -- | ------------ | ---------- |
| ![font-sizes-v2-custom-before](https://github.com/WordPress/gutenberg/assets/5129775/9310f4c1-37c3-42d4-bef8-efe7c7b08e99) | ![font-sizes-v2-custom-after](https://github.com/WordPress/gutenberg/assets/5129775/e8039677-055f-4c9b-8731-4f3480721c25) | ![font-sizes-v3-custom-true](https://github.com/WordPress/gutenberg/assets/5129775/eb3c6c75-cdb9-42e6-b03d-21abee3eb2e3) | ![font-sizes-v3-custom-false](https://github.com/WordPress/gutenberg/assets/5129775/435fb7ff-0b11-4dbc-8b56-8b1f244b9d25) |

### TT4 Empty `fontSizes`

<details>
<summary>Example JSON</summary>

```jsonc
// Empty fontSizes
{
	"settings": {
		"typography": {
			"fontSizes": []
		}
	}
}
```

</details>

| v2 trunk | v2 | v3 (default) | v3 `false` |
| -------- | -- | ------------ | ---------- |
| ![font-sizes-v2-empty-before](https://github.com/WordPress/gutenberg/assets/5129775/bc992dd4-d711-4942-ad32-dcd570ffd2e6) | ![font-sizes-v2-empty-after](https://github.com/WordPress/gutenberg/assets/5129775/6d776bf3-13ab-461f-955e-f6e35f246c8b) | ![font-sizes-v3-empty-true](https://github.com/WordPress/gutenberg/assets/5129775/565c07b7-9c5f-47fa-9395-2532617a32db) | ![font-sizes-v3-empty-false](https://github.com/WordPress/gutenberg/assets/5129775/e613f04f-bf8b-4a93-bf45-686b2deaa80f) |

### TT4 Undefined `fontSizes`

<details>
<summary>Example JSON</summary>

```jsonc
// Undefined fontSizes
{
	"settings": {
		"typography": {}
	}
}
```

</details>

| v2 trunk | v2 | v3 (default) | v3 `false` |
| -------- | -- | ------------ | ---------- |
| ![font-sizes-v2-undefined-before](https://github.com/WordPress/gutenberg/assets/5129775/461deb78-a6e2-4c20-9417-1243f4ed0f44) | ![font-sizes-v2-undefined-after](https://github.com/WordPress/gutenberg/assets/5129775/a300104d-6d22-4ff9-9e91-9f7a69b5059f) | ![font-sizes-v3-undefined-true](https://github.com/WordPress/gutenberg/assets/5129775/2ebde0c6-26d9-44bb-bff1-c11181eafcef) | ![font-sizes-v3-undefined-false](https://github.com/WordPress/gutenberg/assets/5129775/4aa5d285-d605-4edd-9516-2358dc8873cd) |

ajlende committed:
Add defaultFontSizes option to theme.json (#56661)

* Add defaultFontSizes option for theme.json

* Fix defaultFontSizes causing errors when missing

* Simplify useMergedSettings

* Simplify more

* Remove the memo because it's probably fine

* Try ignoring how overrides are handled in the dropdown

* Minor performance improvement using reverse instead of toReversed

* Link to a new issue with a more complete description

* Refactor hack into its own funtion

ajlende committed:
Add migration for theme.json v2 to v3

ajlende committed:
Use 'merge' to mimic old behavior

ajlende committed:
WIP

ajlende committed:
Fix PHP lint

ajlende committed:
Simplify logic

ajlende committed:
Tidy up comments

ajlende committed:
Revert phpunit for now

ajlende committed:
Optimize prevent_override logic

ajlende committed:
preset -> preset_metadata

ajlende committed:
Bump core theme.json version

ajlende committed:
Bump theme.json version in schema

ajlende committed:
Reimplement old logic for "merge" value

ajlende committed:
Remove uniqByProperty

ajlende committed:
Fix WP_Theme_JSON_Gutenberg_Test failures

ajlende committed:
Update WP_Theme_JSON_Schema_Gutenberg_Test with new migration

ajlende committed:
Update migration docs

ajlende committed:
Update phpunit theme fixtures

ajlende committed:
Try fixing WP_REST_Global_Styles_Controller_Gutenberg_Test failure

ajlende committed:
Fix PHP lint

ajlende committed:
Default missing theme.json to latest schema

ajlende committed:
Fix living docs title

ajlende committed:
Refactor typography-panel a bit

ajlende committed:
Fix docs manifest

ajlende committed:
Clarify migration docs

ajlende committed:
Revert "Remove uniqByProperty"

This reverts commit c12d3a28c8b1de32a5ae205d9bcd739e1c8e3958.

ajlende committed:
Fix getUniqueFontSizesBySlug

ajlende committed:
Update default schema elsewhere

ajlende committed:
Order the presets default, theme, custom

ajlende commented on docs/reference-guides/theme-json-reference/theme-json-migrations.md:
```diff
@@ -63,3 +63,23 @@ Additions to styles:
+
+The new `defaultFontSizes` option differs from that behavior.
+
+- When set to `true` it will show the default font sizes and prevent them from being overridden by the theme.
+- When set to `false` it will hide the default font sizes and allow the theme to use the default slugs.
```
I intentionally did not document the `'merge'` option as I don't think theme developers should be using it. It's more of an internal migration thing.

ajlende commented on lib/class-wp-theme-json-data-gutenberg.php:
```diff
@@ -38,7 +38,7 @@ class WP_Theme_JSON_Data_Gutenberg {
 	 * @param string $origin The origin of the data: default, theme, user.
 	 */
-	public function __construct( $data = array(), $origin = 'theme' ) {
+	public function __construct( $data = array( 'version' => WP_Theme_JSON_Gutenberg::LATEST_SCHEMA ), $origin = 'theme' ) {
```
I looked for all the places that define a default theme.json array and added the version to them. Without the version, they run all the migrations which would trigger adding the `'defaultFontSizes' => 'merge'` behavior.

draganescu commented on packages/block-editor/src/components/global-styles/typography-panel.js:
```diff
@@ -102,14 +105,37 @@ function useHasTextColumnsControl( settings ) {
+function getMergedFontSizes( settings ) {
+	/*
+	 * For backwards compatibility with presets converting from a hardcoded `false`
+	 * for `prevent_override` to a path to a boolean (`defaultFontSizes`, for example),
```
`for `prevent_override` to a path to a boolean`

This could be rephrased. Also this comment should say why this is needed (as in what happens without it).

ajlende committed:
Reword the theme.json v2 backwards compatibility comment

ajlende committed:
Update rest controller test theme

ajlende committed:
PHP Lint

ajlende committed:
Update a few more theme.json files for tests

ajlende commented:
I found an issue that I don't have a great solution for. Maybe someone else has an idea.

Any user-configured settings for the theme (such as selecting a different style variation) are saved in the database as a theme.json version 2. This user origin will be passed through the migration causing the page styles to be "dirty" and require a save in the site editor. The diff will look something like this.

```diff
--- a/themev2
+++ b/themev3
@@ -1,6 +1,6 @@
 {
     "$schema": "https://schemas.wp.org/trunk/theme.json",
-    "version": 2,
+    "version": 3,
     "settings": {
         "appearanceTools": false,
         "useRootPaddingAwareAlignments": false,
@@ -236,6 +236,7 @@
         },
         "typography": {
             "customFontSize": true,
+            "defaultFontSizes": "merge",
             "dropCap": true,
             "fontSizes": [
                 {
```

The problem is the `"defaultFontSizes": "merge"` line. Having it saved in the database would leave no way for a user to opt-in to the new default behavior like the theme is able to do by updating the theme.json version.

I think I'm okay with keeping the save button active when in the site editor since we have a similar pattern for block migrations. But I'd like a way for the user to either explicitly opt-in or take the value from the theme if the theme has migrated. i.e. not save `defaultFontSizes` in the database on migration or ignore it when the value is `"merge"` in the user origin.

Maybe someone following here has thoughts/opinions?

draganescu commented:
Can we show this migration as a checklist item in the save dialog which the user could uncheck?

gaambo commented:
I just tested this with [WordPress playground](https://playground.wordpress.net/gutenberg.html).

1. Used TT4 theme
2. Directly edited (yeah, just for the quick tests via playground 😅) TT4s theme.json:
   - replaced the fontSizes with:
```json
"fontSizes": [
        {
          "fluid": false,
          "name": "Sehr klein",
          "size": "0.75rem",
          "slug": "xx-small"
        },
        {
          "fluid": false,
          "name": "Kleiner",
          "size": "0.875rem",
          "slug": "x-small"
        },
        {
          "fluid": false,
          "name": "Klein",
          "size": "1rem",
          "slug": "small"
        },
        {
          "fluid": false,
          "name": "Standard",
          "size": "1.125rem",
          "slug": "standard"
        },
        {
          "fluid": false,
          "name": "Groß",
          "size": "1.375rem",
          "slug": "large"
        },
        {
          "fluid": false,
          "name": "Größer",
          "size": "1.75rem",
          "slug": "x-large"
        },
        {
          "fluid": false,
          "name": "Sehr Groß",
          "size": "2.25rem",
          "slug": "xx-large"
        },
        {
          "fluid": false,
          "name": "Extra Groß",
          "size": "3.25rem",
          "slug": "xxx-large"
        }
      ],
```
- ✅ The font size selector is rendered as a dropdown (because of the amount of fontSizes)
- ✅ Edit a post with a core/paragrah block: The font size selector shows the "small" font size with the correct size (1rem) instead of the default 13px
- 🛑✅ Set the theme.json version to 3 and set `settings.typography.defaultFontSizes` to false
  - The font size picker shows my custom defined fontSizes, but for some reason it still shows the `medium` core default font size
  - On further testing, the reason seams to be because "Medium" is set as the font size for text in the global styles. If I choose another font size there, the font size selector when editing a post correctly only shows the custom font sizes. 

I actually think it makes sense, that any value that is selected on a global style is still also available in the font size selector. 
So imho this works great 👍👏

ajlende committed:
Merge branch 'trunk' into fix/default-font-sizes-override

ajlende committed:
Remove the merge keyword that is no longer needed since #58951

ajlende commented:
> Can we show this migration as a checklist item in the save dialog which the user could uncheck?

This is what you currently see. The behavior of unchecking something in all cases is that it leaves the post unsaved. I wouldn't want unchecking something and hitting save to save the new version but not the changed default when unchecked. It wouldn't be very clear what is going on, and there would be no way to change your mind.

ajlende committed:
Merge branch 'trunk' into fix/default-font-sizes-override

Conflicts:
	packages/block-editor/src/components/global-styles/typography-panel.js

ajlende committed:
Merge branch 'trunk' into fix/default-font-sizes-override

ajlende committed:
Merge branch 'trunk' into fix/default-font-sizes-override

ajlende committed:
Hacky way to fix DB save issues

ajlende commented:
For theme.json font sizes, the `prevent_override` setting has changed from defaulting to `false` to defaulting to `true` because of the addition of a `defaultFontSizes` option in this PR and in order to match the other presets with a `default*` setting in theme.json (ex. `defaultPalette`) which is `true` by default.

The `prevent_override` setting affects the output CSS on the page, so we can't just change the default behavior without adding a new theme.json version.

The [`WP_Theme_JSON_Schema` migration](https://github.com/WordPress/gutenberg/pull/58409/files#diff-abbe46088327fd585719b5e6db810fc1b65104231b47da3bfb9eace32172f6c1R97-R125) I've added works fine for themes. Any theme with `"version": 2` set will default `defaultFontSizes` to `false` which, in turn, defaults the `prevent_override` to `false` keeping the CSS output the same as before. All existing themes continue to work the same as they did before. Themes that update to `"version": 3` get the new default behavior and they have the option to turn `defaultFontSizes` to `false` if they wish.

The problem arises for custom (user) styles. They take precedence over the theme's theme.json, and the same migration still happens. The custom styles are updated to have `defaultFontSizes` set as `false` the next time global styles are saved. If a user has any customizations, they have no way to opt-in to the new `defaultFontSizes` behavior even if the theme they're using has chosen to opt-in.

I added a [fix that I'm not satisfied with](https://github.com/WordPress/gutenberg/pull/58409/files#diff-7590ad3a5619e259acec135b3ebc4a1c3ada47eb34a20fcc068d9c6a3c3a5239R339-R348). But this is how to reproduce the problem without the fix. (b0cff007b6729b452f72dc93db94514b2eec0102).

1. Start with a theme without any customizations. (ex. For TT4, delete the `wp-global-styles-twentytwentyfour` and revisions in the DB.)
2. Add a custom color to the palette.
3. See that the `{"settings":{typography":{"defaultFontSizes":false}}}` is now saved in the DB.
4. Update the theme to v3 by editing the `version` in the theme.json file.
5. See that the DB still has the `defaultFontSizes` as `false`. User has not way to opt-in like the theme could.

We'll have to do this in the future for other `default*` settings that may be added, so I want to make it intuitive for later updates. It shouldn't be something that's easy to forget.

cc @youknowriad if you have any thoughts on the right place to handle this.

youknowriad commented:
> For theme.json font sizes, the prevent_override setting has changed from defaulting to false to defaulting to true because of the addition of a defaultFontSizes option in this PR and in order to match the other presets with a default* setting in theme.json (ex. defaultPalette) which is true by default.

Can you explain more what is `prevent_override` ? and why we had to change its default value?

ajlende commented:
> Can you explain more what is `prevent_override` ?

`prevent_override` is the option in `PRESETS_METADATA` that controls whether custom/theme presets override the default presets with the same slug. [[ref](https://github.com/WordPress/gutenberg/blob/afc08a1ca41dfb4c2e9fd6effd3d9878c1955381/lib/class-wp-theme-json-gutenberg.php#L81-L89)]

For example, let's say a theme defines a `{ "slug": "small", "size": "11px" }` font size preset. With `prevent_override === true`, the output CSS would be `--wp--preset--font-size--small: 13px;` because the default lib/theme.json defines `{ "slug": "small", "size": "13px" }` and we're preventing the default from being overridden. With `prevent_override === false`, you get `--wp--preset--font-size--small: 11px;` in the CSS from the theme's theme.json.

> and why we had to change its default value?

The `prevent_override` option is controlled by the `default*` settings when they are available (ex. `defaultPalette`, `defaultGradient`, `defaultDuotone`, etc.).

The `default*` settings also control whether or not to show the default presets in the UI.

So `"defaultPalette": true` would show the core presets and prevent the theme from overriding them in the generated CSS.

All of the `default*` settings default to `true`, so for consistency I think it would make the most sense to make `defaultFontSizes` default to `true` as well. As a theme developer, I don't want to have to always look up which `default*` setting default to `true` and which ones default to `false`, I would expect them all to be the same.

By setting `defaultFontSizes` to `true` by default, we also end up changing `prevent_override` to `true` by default as well.

youknowriad commented:
@ajlende Thanks for explaining it makes more sense. 

It does seem to be a big code change for a beneficial change (consistency) but we need to question whether the added complexity is worth it. A migration might be fine, but now I'm curious if we should run the migration over the saved theme.json too. Why was the migration not sufficient?

I'm also going to ping @oandregal here as he'll have better insights than me.

ajlende commented:
> Why was the migration not sufficient?

The migration converts any older theme.json version into a `"version": 3` with `"defaultFontSizes": false`.

In the user case, the migration is run whenever the value is read from the DB ([here](https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-rest-global-styles-controller-gutenberg.php#L370)). That migrated data (including the `"version": 3` and `"defaultFontSizes": false`) is then sent back to the DB on save where it can no longer be changed because the migration has already run and there is not UI for the user to go back and change it.

Regardless of the UI, however, I wouldn't want to automatically set this value in the DB because the user hasn't taken any action to say they want this value to override the theme value.

Hopefully that makes sense. Thanks for your help :)

youknowriad commented:
It makes a lot of sense. Thanks @ajlende 

So if I understand properly given:

 - Saved global styles always "override" the theme.json but get merged on top of it.
 - There's no UI to tweak this setting, we can expect that no one actually touched that setting in global styles.

Meaning that if we run the migration only on the files coming from the theme but exclude it from running on the saved theme.json, it will solve the problem no? (maybe won't solve it for style variations but maybe we'll be solving it for most cases which would be enough, how many style variations actually override this setting in existing themes)

Did I miss anything? Do we have a way to know whether a migration is running for theme.json file or saved global styles?

ajlende committed:
Use $new instead of $old

ajlende committed:
Skip defaultFontSizes migration for custom origin theme.json files

ajlende committed:
Merge branch 'trunk' into fix/default-font-sizes-override

ajlende commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2831,12 +2833,15 @@ public function merge( $incoming ) {
 			}
 
 			// Replace the presets.
-			foreach ( static::PRESETS_METADATA as $preset ) {
```
This is named `$preset_metadata` in other places and I think it makes more sense, so I renamed it here.

ajlende commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2859,19 +2865,17 @@ public function merge( $incoming ) {
+					// Filter out default slugs from theme presets when defaults should not be overridden.
+					if ( 'theme' === $origin && $prevent_override ) {
+						$slugs_node    = static::get_default_slugs( $this->theme_json, $node['path'] );
+						$preset_global = _wp_array_get( $slugs_global, $preset_metadata['path'], array() );
+						$preset_node   = _wp_array_get( $slugs_node, $preset_metadata['path'], array() );
+						$preset_slugs  = array_merge_recursive( $preset_global, $preset_node );
 
-						$slugs_for_preset = _wp_array_get( $slugs, $preset['path'], array() );
-						$content          = static::filter_slugs( $content, $slugs_for_preset );
-						_wp_array_set( $this->theme_json, $path, $content );
+						$content = static::filter_slugs( $content, $preset_slugs );
 					}
+
+					_wp_array_set( $this->theme_json, $path, $content );
```
This is just a refactor so things look a little more clear. This is the process that I did to make sure the logic for the conditional was exactly the same.

```php
// Initial NOT (!) because of else clause
// $override_preset === ! $prevent_override

! ( ( 'theme' !== $origin ) || ( 'theme' === $origin && ! $prevent_override ) )

! ( 'theme' !== $origin ) && ! ( 'theme' === $origin && ! $prevent_override )

'theme' === $origin && ! ( 'theme' === $origin && ! $prevent_override )

'theme' === $origin && ( 'theme' !== $origin || $prevent_override )

( 'theme' === $origin && 'theme' !== $origin ) || ( 'theme' === $origin && $prevent_override )

false || ( 'theme' === $origin && $prevent_override )

'theme' === $origin && $prevent_override
```

Both code paths ended in `_wp_array_set( $this->theme_json, $path, $content );` so that was moved out.

ajlende commented on lib/class-wp-theme-json-resolver-gutenberg.php:
```diff
@@ -244,7 +244,7 @@ public static function get_theme_data( $deprecated = array(), $options = array()
 				$theme_json_data = static::read_json_file( $theme_json_file );
 				$theme_json_data = static::translate( $theme_json_data, $wp_theme->get( 'TextDomain' ) );
 			} else {
-				$theme_json_data = array();
```
An empty array is equivalent to an empty v1 theme.json as far as migrations are concerned. I went through all the code and looked for places like this where we want to start with an empty latest version theme.json instead of v1.

ajlende commented on phpunit/data/themedir1/block-theme-child-with-fluid-layout/theme.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"$schema": "https://schemas.wp.org/trunk/theme.json",
-	"version": 2,
```
I don't think there's a good way to avoid having to update the theme.json version in all of the fixtures.

Many of the tests use deep equivalence to check what they are testing which means the addition of `"defaultFontSizes": false` breaks those tests even though that's not what they're testing. Rather than going through and fixing all the tests, it seemed easier to just update the fixtures.

ajlende reviewed (commented):
This is ready for review. The custom settings migration seems to be working now and I added a few comments to try and explain why I made some of the changes.

ajlende commented:
> One thing I've noticed is that I cannot reproduce https://github.com/WordPress/gutenberg/issues/52200 the issue we aim to fix. . . . Isn't this what we aim to achieve with this PR? I may be missing other context.

There are a lot of related issues and a couple reverted PRs in the mix, but I'll try to summarize the context and provide links.

The original issue was more an issue of order for the presets which was also reported in https://github.com/WordPress/gutenberg/issues/55744. I suggested adding the `defaultFontSizes` option to make this consistent with other presets https://github.com/WordPress/gutenberg/issues/52200#issuecomment-1828143458 as a more complete solution.

I tried once with https://github.com/WordPress/gutenberg/pull/56661 but ended up breaking back compat, so that was reverted https://github.com/WordPress/gutenberg/pull/58456.

While working on that, I also found an issue with style generation for font sizes https://github.com/WordPress/gutenberg/issues/57733.

Then, the PR that added the merged font sizes https://github.com/WordPress/gutenberg/pull/55219 ended up being reverted too https://github.com/WordPress/gutenberg/pull/58951. The intention for that PR was to show default font sizes which still needed to be done again. So all that was left was adding the `defaultFontSizes` which was only really being tracked by #52200.

And that brings us here to this PR that should wrap up the loose ends from all of that. 😅

ajlende commented:
> I've noticed the UI control for font sizes is quite particular. It takes the font sizes defined in the theme's theme.json and creates a S-M-L-... visual scale in the UI control. It ignores font sizes names (only displayed as labels when you hover over it). And the S-M-L values are taken as they are sorted in theme.json, no matter their actual size.

Yeah, it is a bit of an unexpected implementation. I believe that behavior existed before any of the PRs/issues mentioned here.

> If we relax this setup would theme switching for font sizes become more of an issue?

I believe the underlying styles all still use the slugs, so it's really just a problem with the UI "lying" about the sizes.

I can fix that in this PR if you think it needs it to be complete, but I'd prefer to leave it as a follow-up. I'm not sure what the most clear way for aligning presets to t-shirt sizes would be.

gaambo commented:
> A more abstract note. While I was checking how font sizes work, I've noticed the UI control for font sizes is quite particular. It takes the font sizes defined in the theme's theme.json and creates a S-M-L-... visual scale in the UI control. It ignores font sizes names (only displayed as labels when you hover over it). And the S-M-L values are taken as they are sorted in `theme.json`, no matter their actual size. For example, I've set `version:3` and `defaultFontSizes: false` and then provided two fonts:

I think this is only a UI thing. As far as I know, the UI for typography presets displays buttons with the S-M-L labels if there are less than 7 preset sizes. For more than 7 a select is displayed with the correct labels from theme.json: 
![grafik](https://github.com/WordPress/gutenberg/assets/5585580/a7410503-e850-4737-9811-ff378e2ec352)

But it always uses the font sizes slug for generating the class names. I guess the S-M-L is just an abbreviated label for the buttons. I also think that some time in the past "1-2-3" were being used as labels.

So I would suggest leaving that out of this PR 😅

oandregal commented:
Thanks for all the context, Alex!

So, these were all issues that we had:

- https://github.com/WordPress/gutenberg/issues/57733
- https://github.com/WordPress/gutenberg/issues/55744
- https://github.com/WordPress/gutenberg/issues/52200

I cannot reproduce any of them using TwentyTwentyFour, WordPress 6.5, with and without Gutenberg `trunk` active.

ajlende commented:
> I cannot reproduce any of them using TwentyTwentyFour, WordPress 6.5, with and without Gutenberg trunk active.

Thanks @oandregal for checking, I suppose we can close them then and update the description. However, I think that this PR is still valid for bringing some consistency to configuration options in theme.json.

As a follow-up to this, it wouldn't be too hard to add the options for `fontFamilies` and `spacingSizes` to get them all caught up with options that I would expect to have. We could squeeze those into the same theme.json version bump.

scruffian commented:
> 5. Update your theme.json (again) with settings.typography.defaultFontSizes to true.
> 6. See that the default values do not show and the theme values are not overridden.

Are these instructions correct? I think it should say update `defaultFontSizes` to false (since true is the default)...

Also, when I'm testing theme.json v2,  I don't see the default sizes even if I rename `small` to `footnote`.

ajlende committed:
Merge branch 'trunk' into fix/default-font-sizes-override

Conflicts:
	lib/class-wp-theme-json-gutenberg.php

ajlende committed:
Fix missing textAlign in docs from merge

ajlende committed:
Remove empty line change

ajlende committed:
Fix user styles not migrating correctly

ajlende committed:
Reorder font size dropdown

ajlende committed:
Only run migration on themes with font sizes defined

ajlende committed:
Add more details to comment

ajlende commented:
> > 5. Update your theme.json (again) with settings.typography.defaultFontSizes to true.
> > 6. See that the default values do not show and the theme values are not overridden.
> 
> Are these instructions correct? I think it should say update `defaultFontSizes` to false (since true is the default)...
> 
> Also, when I'm testing theme.json v2, I don't see the default sizes even if I rename `small` to `footnote`.

@scruffian It looks like I missed updating some parts over the course of this being open while other PRs were merged and old behavior was changed. I've updated the screenshots and instructions to hopefully be more clear. Thanks for reviewing!

ajlende committed:
Ensure isGlobalStylesUserThemeJSON is set for migrations

ajlende commented on lib/class-wp-theme-json-resolver-gutenberg.php:
```diff
@@ -549,14 +549,17 @@ public static function get_user_data() {
+		// Needs to be set for schema migrations of user data.
+		$config['isGlobalStylesUserThemeJSON'] = true;
+
 		static::$user = new WP_Theme_JSON_Gutenberg( $config, 'custom' );
```
`isGlobalStylesUserThemeJSON` needs to be preset for migrations to run. Migrations are run on both `WP_Theme_JSON_Data_Gutenberg` and `WP_Theme_JSON_Gutenberg` constructors.

scruffian reviewed (approved):
I have tested all of the scenarios and they all work as described. The code makes sense to me too.

ajlende committed:
Update tests since skipping for missing font sizes

ajlende committed:
Fix phpcs

ajlende committed:
Update migration docs

ajlende committed:
Merge branch 'trunk' into fix/default-font-sizes-override

ajlende commented:
Dev note draft 📝 

---

## Theme.json version 3

The theme.json version has been incremented in order to update the default behavior for overriding the default `fontSizes` presets in theme.json.

Using older theme.json versions will continue to be supported and should have no change. However, upgrading to version 3 is recommended because new development will continue in the newer versions.

See the [theme.json reference](https://developer.wordpress.org/block-editor/reference-guides/theme-json-reference/theme-json-migrations/#migrating-from-v2-to-v3) for more info on migration.

### `defaultFontSizes`

A new `settings.typography.defaultFontSizes` option has been added to give control over showing default (core-provided) font sizes and preventing those defaults from being overridden.

- When set to `true` it will show the default font sizes and prevent them from being overridden by the theme.
- When set to `false` it will hide the default font sizes and allow the theme to use the default slugs.

It is `true` by default when switching to v3. This is to be consistent with how other `default*` options work such as `settings.color.defaultPalette`, but differs from the behavior in v2.

In theme.json v2, the default font sizes were only shown when theme sizes were not defined. A theme providing font sizes with the same slugs as the defaults would always override the default ones.

To keep behavior similar to v2, set `settings.typography.defaultFontSizes` to `false`. For example:

```json
{
	"settings": {
		"typography": {
			"defaultFontSizes": false,
			"fontSizes": [
				{
					"name": "Small",
					"slug": "small",
					"size": "10px"
				},
				{
					"name": "Medium",
					"slug": "small",
					"size": "16px"
				},
				{
					"name": "Large",
					"slug": "large",
					"size": "28px"
				},
			]
		}
	}
}
```

carolinan commented:
@ajlende The documentation updates look mostly good but there are some files that were not updated, and broken links.

Needs updating:
- The menu: reference-guides/toc.json
- The theme-json-reference/README.md file: https://developer.wordpress.org/block-editor/reference-guides/theme-json-reference/
- theme-json-reference/theme-json-migrations.md has a heading that is incorrect:
```
## Migrating from v2 to v3

Upgrading to v3 adjusts preset defaults to be more consistent with one another.

### How to migrate from v1 to v2:
```

t-hamano commented:
JSON Schema no longer allows version 2. Shouldn't version 2 still be allowed?

![image](https://github.com/WordPress/gutenberg/assets/54422211/08ab47c3-4b7b-4972-9145-bc1003ca4be8)

ajlende commented:
> JSON Schema no longer allows version 2. Shouldn't version 2 still be allowed?

While v2 theme.json files will continue to work on current versions of WordPress, the schema is a tool for theme developers to verify their theme.json with the specified version of WordPress you are developing for. Versions of schemas are handled via URL. For example, the last theme.json v2 schema will be located at `https://schemas.wp.org/wp/6.5/theme.json`, and the last theme.json v1 schema is at `https://schemas.wp.org/wp/5.8/theme.json`.

I didn't do a great job of explaining this in the [original theme.json blog post](https://make.wordpress.org/themes/2021/11/30/theme-json-schema/#versioning), but unless you are developing for folks with the Gutenberg plugin active, you don't really want to be using the `trunk` version of the schema. A theme should be using the oldest version of WordPress that it supports.

Technically JSON Schema Draft-04 would be able to support all versions simultaneously, but it would be a nightmare to maintain, would be slower for code editors to check, and it wouldn't flag developers working off trunk of a new version where all new work is happening.

ajlende commented:
> The documentation updates look mostly good but there are some files that were not updated, and broken links.

@carolinan Thanks for catching those documentation issues. I found a bunch more issues as I was fixing those things and added everything in https://github.com/WordPress/gutenberg/pull/61221.

Mamaduka commented on schemas/json/theme.json:
```diff
@@ -2244,7 +2249,7 @@
 			"description": "Version of theme.json to use.",
 			"type": "integer",
-			"enum": [ 2 ]
+			"enum": [ 3 ]
```
I think we still need to support the older versions. Having only version 3 in the enum list produces a warning for default themes.

![CleanShot 2024-04-30 at 11 33 35](https://github.com/WordPress/gutenberg/assets/240569/bb37ea3b-17c6-4f6b-822b-c460fd7210a0)

t-hamano replied:
Or should we increase the default theme's theme.json version to 3?

Mamaduka replied:
I mentioned the default theme as an example. Not everyone will upgrade `versions` immediately, and the older versions still work.

ajlende replied:
Default themes, like all other themes, should be using the schema for the oldest version of WordPress that they intend to support so that incompatible styles and settings are properly flagged. If the theme.json version is updated to 3, then the theme will only work with 6.6 and up or with the latest Gutenberg plugin installed.

It's confused often enough that I'm planning on writing a developer.wordpress.org blog post to help explain how JSON schema versioning works and how it's different from the theme.json version.

fabiankaegy commented on lib/class-wp-theme-json-schema-gutenberg.php:
```diff
@@ -38,6 +38,7 @@ class WP_Theme_JSON_Schema_Gutenberg {
 	 * Function that migrates a given theme.json structure to the last version.
 	 *
 	 * @since 5.9.0
+	 * @since 6.5.0 Migrate up to v3.
```
@ajlende shouldn't this be 6.6?

ajlende replied:
Yep. Updating in #61328

fabiankaegy commented on lib/class-wp-theme-json-schema-gutenberg.php:
```diff
@@ -87,6 +94,56 @@ private static function migrate_v1_to_v2( $old ) {
+	 *
+	 * - Sets settings.typography.defaultFontSizes to false.
+	 *
+	 * @since 6.5.0
```
Again I think this should be 6.6
